src/Selenium2Driver.php

```diff
@@ -758,170 +695,135 @@ public function selectOption($xpath, $value, $multiple = false)
     {
-        try {
-            // Move the mouse to the element as Selenium does not allow clicking on an element which is outside the viewport
-            $this->wdSession->moveto(array('element' => $element->getID()));
-        } catch (UnknownCommand $e) {
-            // If the Webdriver implementation does not support moveto (which is not part of the W3C WebDriver spec), proceed to the click
-        }
-
-        $element->click();
+        // Move the mouse to the element as Selenium does not allow clicking on an element which is outside the viewport
+        $this->webDriver->action()->click($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function doubleClick($xpath)
     {
-        $this->mouseOver($xpath);
-        $this->wdSession->doubleclick();
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->doubleClick($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function rightClick($xpath)
     {
-        $this->mouseOver($xpath);
-        $this->wdSession->click(array('button' => 2));
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->contextClick($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function attachFile($xpath, $path)
     {
+        /** @var RemoteWebElement $element */
         $element = $this->findElement($xpath);
         $this->ensureInputType($element, $xpath, 'file', 'attach a file on');
 
-        // Upload the file to Selenium and use the remote path. This will
-        // ensure that Selenium always has access to the file, even if it runs
-        // as a remote instance.
-        try {
-          $remotePath = $this->uploadFile($path);
-        } catch (\Exception $e) {
-          // File could not be uploaded to remote instance. Use the local path.
-          $remotePath = $path;
-        }
-
-        $element->postValue(array('value' => array($remotePath)));
+        $element->setFileDetector(new LocalFileDetector());
+        return $element->sendKeys($path);
     }
 
     /**
      * {@inheritdoc}
      */
     public function isVisible($xpath)
     {
-        return $this->findElement($xpath)->displayed();
+        $element = $this->findElement($xpath);
+        return $element->isDisplayed();
     }
 
     /**
      * {@inheritdoc}
      */
     public function mouseOver($xpath)
     {
-        $this->wdSession->moveto(array(
-            'element' => $this->findElement($xpath)->getID()
-        ));
+        $element = $this->findElement($xpath);
+        $this->webDriver->action()->moveToElement($element)->perform();
     }
 
     /**
      * {@inheritdoc}
      */
     public function focus($xpath)
     {
-        $this->trigger($xpath, 'focus');
+        $element = $this->findElement($xpath);
+        $action = $this->webDriver->action();
+
+        $action->moveToElement($element);
+        $action->click($element);
```

Comment:
clicking an element is not the same than focusing it. It triggers extra actions (for a link, it might navigate for instance)

Reply:
will investigate, good catch

Reply:
fixed